### DISCUSSION TITLE
Question: What is the fastest method to draw thousands of polygons?

### DISCUSSION THREAD

**Discussion by feclaudia** [Mapsui/Mapsui#1867] [Q&A]
Hello!

My task is to draw a few thousands polygons (with an OSM background) in one image and change colours of all polygons (but not the geometries) for the following images. My methods are to slow....

I used a list of polygons:
`List<IFeature> PolygonList = new List<IFeature>();`

I added polygons with:
`var Polygon = new GeometryFeature { [Name] = ii + " " + jj, Geometry = new NetTopologySuite.Geometries.Polygon(new NetTopologySuite.Geometries.LinearRing(new[] { new Coordinate(ms1.X, ms1.Y), new Coordinate(ms2.X, ms2.Y), new Coordinate(ms3.X, ms3.Y), new Coordinate(ms4.X, ms4.Y), new Coordinate(ms1.X, ms1.Y) })) };`                         
`Polygon.Styles.Add(mscolor);`
`PolygonList.Add(Polygon);`

I added the new layer with:
`MyMapControl.Map.Layers.Add(new MemoryLayer()
{
      Style = null,
      Features = PolygonList,
      Name = "Data",
      IsMapInfoLayer = true
});`

For the following images I took the styles of the PolygonList and changed the colour:
`msco = PolygonList[(ii * zeitreihe[zIndex].nrows) + jj].Styles.First() as VectorStyle;`
`msco.Fill.Color = msc;`

Then I updated my image with:
`MyMapControl.Map.Layers[dl].DataHasChanged();`

I hoped, that my image display was getting faster with the new geometry structures, but it was slower. So there must be another solution. Do you have an idea or maybe an example for me?

I'm thankful for any advice!
Thanks, Claudia

**Comment by pauldendulk** [comment]
In the samples there is a 'Performance' category with some possible improvements. You could wrap your layer in a RasterizingLayer or a RasterizingTileLayer.

**Reply by feclaudia** [reply]
I tried ro wrap my layer, but with the RasterizingTileLayer the DataHasChanged didn't work. With RasterizingLayer the DataHasChanged  worked, but still to slow. Probably I'm not using and updating the RasterizingLayer correctly..... I have no experience with that.

**Comment by inforithmics** [comment]
I would use a GemetrySimplifyProvider combined with a Rasterizing Tile Layer. Because I think to many points are drawn. The GeometrSimplifyProvider reduces the geometric Complexity at higher Zoom levels and therefore improving Performance.

**Reply by feclaudia** [reply]
That sounds interesting! Is there an example in the samples, how to use GeometrySimplifyProvider?

**Reply by feclaudia** [reply]
I just found the GeometrySimplifyProvider in
Samples/Mapsui.Samples.Common/Maps/Performance/GeometrySimplifySample.cs

**Reply by feclaudia** [reply]
I tried the  GeometrySimplifyProvider with 
`var smemo = new MemoryProvider(PolygonList);`
`var simplifyPolygonList = new GeometrySimplifyProvider(smemo, distanceTolerance: 500000);`
and tried different values of distanceTolerance, but could't see a difference. What unit is this?

**Reply by thomasstockermc** [reply]
This is the Resolution (How many units a Pixel contains). Not setting this calculates this automatically with the Zoom Level. This should be used I think. The fixed value is only to display the effect of the reduced geometry.

**Comment by inforithmics** [comment]
I would combine it with GeometryIntersectionProvider which removes the Geometry Areas outsize of the visible area. So that less points are rendered or used.

https://github.com/Mapsui/Mapsui/pull/1820

Actually there is a VectorCache in Mapsui with a LRU size of 10000 Elements, when you have more Elements you need to increase the number of cached Elements. Currently when you zoom or move the Elements get recalculated this needs to be improved. But on a Static map the Geometries are cached.

public VectorCache(ISymbolCache symbolCache, int capacity)

usage:

mapRenderer.RenderCache.VectorCache = new VectorCache(mapRendererWithoutCache.RenderCache.SymbolCache, 100000);

**Reply by feclaudia** [reply]
I'm not sure, how to implement this. The GeometryIntersectionProvider is not included in my Mapsui.Nts.Providers. (I have Mapsui 4.0.0-beta.5). There is just the GeometrySimplifyProvider included.

**Comment by inforithmics** [comment]
The geometry intersection provider was introduced in a later mapsui version. Is there a reason why beta 5 is used and not the newest version?

**Reply by feclaudia** [reply]
I had framework version problems..... I started yesterday a new test project, where I have the newest version. I will try all the different advices there....

**Reply by feclaudia** [reply]
Please help me with VectorCache and the GeometryIntersectionProvider. This is my test program now, where I test the normal and the simplified layer:

`public MainWindow()
        {
            InitializeComponent();
            var mapControl = new Mapsui.UI.Wpf.MapControl();
            mapControl.Map?.Layers.Add(Mapsui.Tiling.OpenStreetMap.CreateTileLayer());

            var layer_normal = CreateLayer();
            //var layer_simplify = SimpleCreateLayer();
            mapControl.Map?.Layers.Add(layer_normal);
            //mapControl.Map?.Layers.Add(layer_simplify);
            //mapControl.Map?.Layers.Add(new RasterizingLayer(layer_normal));
            //mapControl.Map?.Layers.Add(new RasterizingLayer(layer_simplify));

            Content = mapControl;
        }
        
        public static ILayer SimpleCreateLayer()
        {
            return new Layer("Polygons")
            {
                DataSource = new Mapsui.Nts.Providers.GeometrySimplifyProvider(new MemoryProvider(CreatePolygon().ToFeatures())),
                Style = new VectorStyle
                {
                    Fill = new Mapsui.Styles.Brush(Mapsui.Styles.Color.Red),
                }
            };
        }

        public static ILayer CreateLayer()
        {
            return new Layer("Polygons")
            {
                DataSource = new MemoryProvider(CreatePolygon().ToFeatures()),
                Style = new VectorStyle
                {
                    Fill = new Mapsui.Styles.Brush(Mapsui.Styles.Color.Red),
                }
            };
        }
        private static List<NetTopologySuite.Geometries.Polygon> CreatePolygon()
        {
            var result = new List<NetTopologySuite.Geometries.Polygon>();

            NetTopologySuite.Geometries.Polygon polygon1;
            int factor = 0;

            for (int i = 0; i < 900000; i++)
            {
                factor = i - 100*(int)Math.Round((double)(i / 100));
                polygon1 = new NetTopologySuite.Geometries.Polygon(
                new LinearRing(new[] {
                new Coordinate(1000*(factor-1), 1000*(factor-1)-(Math.Round((double)(i/100))*1000)),
                new Coordinate(1000*(factor-1), 1000*(factor)-(Math.Round((double)(i/100))*1000)),
                new Coordinate(1000*(factor), 1000*(factor)-(Math.Round((double)(i/100))*1000)),
                new Coordinate(1000*(factor), 1000*(factor-1)-(Math.Round((double)(i/100))*1000)),
                new Coordinate(1000*(factor-1), 1000*(factor-1)-(Math.Round((double)(i/100))*1000))
                }));

                result.Add(polygon1);
            }
            return result;
        }`

Do I have to set mapControl.Renderer.RenderCache.VectorCache ? How?
Maybe you can adjust my test in the way, you would do it. Thanks a lot!

**Comment by inforithmics** [comment]
1) I think when there are really 900000 polygons, there needs to be made some improvements in Mapsui for example. Using a Topological Index like. 

https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Index.Strtree.html

This needs to be added to MemoryLayer/MemoryProvider and GeoJsonProvider.

2) So setting the VectorCache would be in your code.

```
var renderCache = mapControl.Renderer.RenderCache;
renderCache.VectorCache = new VectorCache(renderCache.SymbolCache, 1000000);
```

3) I would use RasterizingTileLayer instead of RasterizingLayer because with zoom and pan RasterizingTileLayer is much faster.

**Reply by feclaudia** [reply]
o.k. thanks

**Comment by feclaudia** [comment]
Hey! I'm back to my old polygon problem. 
I had a look on the topolgical index in the nettopologysuite, but how can I use it for the polygon layer? You wrote, it has to be added to MemoryLayer/MemoryProvider and GeoJsonProvider.....
Thanks, Claudia

**Comment by inforithmics** [comment]
You could use the IndexedMemoryProvilder because this uses an Index to fetch the Polygons.

Usage can be seen here
https://github.com/Mapsui/Mapsui/blob/master/Samples/Mapsui.Samples.Common/Maps/Performance/RasterizingTileLayerWithThousandsOfPolygons.cs

**Reply by feclaudia** [reply]
o.k. thank you very much! I will test it in my application and let you know.

**Reply by feclaudia** [reply]
Hello! In my application the polygons have different colours. I did't see, how I can give a NetTopologySuite.Geometries.Polygon a style, so I used GeometryFeature with Polygon as Geometry. I put these GeometryFeatures in the List<IFeature> PolygonList. Then I used the  IndexedMemoryProvider in

var layer_normal = new Layer("Data")
  {
      DataSource = new IndexedMemoryProvider(PolygonList),
      Style = new VectorStyle
      {
          Opacity = (float) 0.2,
          Outline = null,
          Line = null
      }
  };

and added it as RasterizingTileLayer to my map. So far so good. 
The next step is to change the colours of the polygons (not the positions). What is the best and fastest way therefore? At the moment I change the colours in the Polygonlist and try Refresh, RefreshData, RefreshGraphic and/or DataHasChanged, but it does not change the displayed image. It seems, that I missed something....
Thanks, Claudia

**Comment by charlenni** [comment]
Did you add a IStyle to the Styles collection, so that Mapsui knows which color do you like to have?

**Reply by feclaudia** [reply]
I added the colour in a IStyle for the GeometryFeature.

**Comment by charlenni** [comment]
Ah, when pressing the comment button, I saw, that you add a style to the layer. With this, all polygons will have the same style and so the same color. In this case, you should omit the style for the layer and add an individual style to each feature.

**Reply by feclaudia** [reply]
I put the opacity in the layer style.

**Reply by feclaudia** [reply]
If you use the layer colour (one colour for all polygons), the "reload" behaviour after changing the colour is the same. It just reloads after changing zoom.... I tried as well the RasterizingLayer, which has nearly the same  behaviour. It reacts on zoom and click.

**Comment by charlenni** [comment]
So your problem is, that the colors aren't updated? Or that the color of each polygon isn't correct?

**Reply by feclaudia** [reply]
The colours aren't updated.

**Comment by charlenni** [comment]
Additionally, keep in mind, that if you use a layer style and a feature style, each polygon is drawn twice. One time for the layer style (see [here](https://github.com/Mapsui/Mapsui/blob/31b9099c758f93daa5aded630e4ac45ec4308cab/Mapsui/Rendering/VisibleFeatureIterator.cs#L38)) and, after that, one time for each feature (see [here](https://github.com/Mapsui/Mapsui/blob/31b9099c758f93daa5aded630e4ac45ec4308cab/Mapsui/Rendering/VisibleFeatureIterator.cs#L58)). If you set a layer style, it could be useless, because you draw the same feature with a different style. That is only useful, if you want to draw the most features with the same style and only a few ones with a different style.

**Reply by feclaudia** [reply]
o.k.. I deleted the layer style and I'm just using the feature style.

**Comment by charlenni** [comment]
If you want to redraw the RasterizingLayer, then you should call `Map.Refresh()` or `Map.RefrashData()`, because your data has changed. If you call only `Map.RefreshGraphics()`, then the RasterizingLayer doesn't change the data and displays the same rasterized images than before.

**Reply by feclaudia** [reply]
I tried Map.Refresh() and Map.RefrashData(), but no update. I changed the colours in the PolygonList and call then
MyMapControl.Map.Layers[dl].DataHasChanged();
MyMapControl.Map.RefreshData(); or MyMapControl.Map.Refresh()

For testing purpose I added a button for colour change (blue/red) in the simple first example with one colour (in the layer style) for all polygons:

private void farbe_Click(object sender, RoutedEventArgs e)
        {
            var layer = mapControl.Map.Layers[1] as RasterizingTileLayer;
            var colour = layer.SourceLayer.Style as VectorStyle;
            if (wechsel == true)
            {
                colour.Fill = new Mapsui.Styles.Brush(Mapsui.Styles.Color.Blue);
                wechsel = false;
            }
            else if(wechsel == false)
            {
                colour.Fill = new Mapsui.Styles.Brush(Mapsui.Styles.Color.Red);
                wechsel = true;
            }
            mapControl.Map.Layers[1].DataHasChanged();
            mapControl.Map.Refresh();
            //mapControl.Map.RefreshGraphics();
            //mapControl.Map.RefreshData();
            //mapControl.RefreshData();
            //mapControl.ForceUpdate();
            //mapControl.RefreshGraphics();
            //mapControl.Refresh();
        }

The update behaviour is the same as in my application with the PolygonList. If I include the Layer as normal Layer
mapControl.Map?.Layers.Add(layer_normal);

and not as RasterzingTile Layer
mapControl.Map?.Layers.Add(new RasterizingTileLayer(layer_normal));
... it works.

**Comment by charlenni** [comment]
@pauldendulk @inforithmics Tried to look into this. The problem seems, that the RasterizingTileLayer doesn't get an information about data changes. So one first thing should be to add the following line to the constructor of RasterizingTileLayer:

        SourceLayer.DataChanged += (s, e) => DataHasChanged();

And as a second thing in this case, RasterTileLayer has to refresh the cache.

**Reply by feclaudia** [reply]
Anything I can do? Any news? Thanks.

**Comment by inforithmics** [comment]
@charlenni Thanks for investigating this.

**Comment by inforithmics** [comment]
RasterizingTileLayer.ClearCache should help. After changing the color

**Reply by feclaudia** [reply]
Thanks! ClearCache does help to update the colours, but the transition from one image to the next isn't smooth. ClearCache reloads the image, so that a kind of blinking is generated: first image - nothing - second image - nothing - third image.... If I load the image data to quick, tiles of the rastertilelayer are missing.
![example1](https://github.com/Mapsui/Mapsui/assets/102292078/6668fc0b-f4ce-437c-99fc-b89d66aa96af)
![example2](https://github.com/Mapsui/Mapsui/assets/102292078/ca5e6e9f-35b3-4e4b-894b-7f8131130cfb)

**Comment by inforithmics** [comment]
pull request, I plan to add a sample

https://github.com/Mapsui/Mapsui/pull/2106

**Reply by inforithmics** [reply]
I think the ClearCache could be made smarter so that only the tiles that are affected by the change are updated

**Reply by feclaudia** [reply]
o.k. that would be great :-)

**Reply by inforithmics** [reply]
In the pull request I added a TilingRenderFetchStrategy

```
﻿using System.Collections.Generic;
using System.Linq;
using BruTile;
using BruTile.Cache;
using Mapsui.Tiling.Extensions;

namespace Mapsui.Tiling.Rendering;

public class TilingRenderFetchStrategy : IRenderFetchStrategy
{
    private readonly IRenderFetchStrategy _renderFetchStrategy;
    private IList<IFeature>? _lastFetch;

    public TilingRenderFetchStrategy(IRenderFetchStrategy? renderFetchStrategy)
    {
        _renderFetchStrategy = renderFetchStrategy ?? new RenderFetchStrategy();
    }

    public IList<IFeature> Get(MRect extent, double resolution, ITileSchema schema, ITileCache<IFeature?> memoryCache)
    {
        var result = _renderFetchStrategy.Get(extent, resolution, schema, memoryCache);
        FillMissingFeatures(result, extent, resolution, schema);
        _lastFetch = result;
        return result;
    }

    private void FillMissingFeatures(IList<IFeature> result, MRect extent, double resolution, ITileSchema schema)
    {
        var tiles = schema.GetTileInfos(extent.ToExtent(), resolution);
        var missingFeatures = new List<IFeature>();
        foreach (var tileInfo in tiles)
        {
            var mRect = tileInfo.Extent.ToMRect();
            if (!result.Any(f => mRect.Equals(f.Extent)))
            {
                var missingTile = _lastFetch?.FirstOrDefault(f =>  mRect.Equals(f.Extent));
                if (missingTile != null)
                {
                    missingFeatures.Add(missingTile);
                }
            }
        }

        foreach (var tile in missingFeatures)
        {
            result.Add(tile);
        }
    }
}
```

Which can be set as the IRenderFetchStrategy, This strategy works the following way, If a Tile is not loaded or rendered yet it uses an old Tile, so that no holes are rendered. In the sample the Tile Flickering disappeared with that.

But I think for this Use Case 
https://github.com/Mapsui/Mapsui/pull/2124
Is the better approach.

**Reply by feclaudia** [reply]
Thanks! I will test this as well.

**Comment by charlenni** [comment]
When looking on your images: How do you get the data for the polygons (format)? I'm not sure, if drawing polygons in this case is the best. You have surely cells with data, that are squares of a given size. Wouldn't it be better to create an image, which each pixel corresponds to a cell, which you draw over the map?

**Reply by feclaudia** [reply]
Hey! I have polar and Cartesian Data, which I want to display. My polygons are for one dataset always the same, but the colours have to change form image to image quickly (based on rain data). And I have to show the values whith mouseover, which is difficult with an image ( I would think). How would a solution with an image looks like?

**Reply by charlenni** [reply]
Sorry, but you didn't answered the question: How big are your squares you have? 100 m², 500 m² or 2 km²? And another one: you have always the same part of the map (static) and want to display a "movie" how the rain is moving, correct?

**Reply by feclaudia** [reply]
Sorry, if I didn't get you right. The size of the square is around 1km². For polar data I have the point in the middle and then a grid of 1° x 1km or 1° x 250m for instance depending on the data. For Cartesian data I have the corner points and the grid is e.g.1km x 1km. For all data I calculate the right position of my cells on the earth surface and draw the polygons. The map is the same for one dataset, but it has to be zoom proof. The rain should be displayed as "movie", correct.

**Reply by charlenni** [reply]
My idea would be to create a bitmap (with Skia?). Each pixel correspondents to a square of your data. E.g. you have a 100 x 100 km^2 area of data with a resolution of 1 x 1 km^2. Then create a bitmap with 100 x 100 pixels, each containing a color corresponding to your data. This bitmap is used for a RasterFeature. With the given Extend of the data, this bitmap is drawn each time over the map if you zoom the map, the bitmap is also zoomed. Each pixel is resized to rect of the correct size. This should be much faster then to redraw all the polygons, because Skia is optimized to draw resized bitmaps.

And for making a movie, you need only more than one RasterFeature.

**Reply by feclaudia** [reply]
Does this also work for polar data? Then the pixels have to look like pieces of cake....

**Reply by charlenni** [reply]
It is a mater of how you draw your bitmap. In this case, you would use a larger bitmap. It isn't so easy as with the square data, but should be possible. The idea behind it is, that you draw your data only once and then Skia doesn't have to redraw it again and again (with all the conversions from Feature to SKPath), but need only to resize a bitmap.

**Reply by feclaudia** [reply]
Sounds good so far. I will try to implement it. Does an example exist, which I can start with? Thanks.

**Reply by charlenni** [reply]
I made a small demo (#2124) to show, how I think, that should work.

There are two different demos in one: a bitmap demo for cartesian data and a polar data bitmap drawn with SkiaSharp commands. The second way could also be used for cartesian data. You will see pixels, if you zoom in too much. Could be changed by increasing size of bitmap (not to high, because of speed reasons), but you said, your zoom is fixed.

**Reply by feclaudia** [reply]
Thanks a lot! I will test it. Zoom is not fixed.....

**Reply by feclaudia** [reply]
![png_pixel](https://github.com/Mapsui/Mapsui/assets/102292078/121970bb-0f7e-4e0c-a097-b64e462659af)
In my first try with cartesian data I cannot zoom in properly. Maybe I missed something for resizing the bitmap? Thanks.

**Reply by charlenni** [reply]
No, that is ok. First, you will have different data. Your data has not only single points, but consists of clusters of the same points. Second, you could increase the size of the created image. In the case of the demo, only a 250 pixel x 250 pixel bitmap is used. You could increase this, but then you should use the unsafe version.

**Reply by feclaudia** [reply]
o.k. thanks

**Reply by charlenni** [reply]
If you create your own Style and StyleRenderer, you could get this (what I assume, that it is more, what you expected)

![grafik](https://github.com/Mapsui/Mapsui/assets/4381223/0eeaa6ae-d5df-4141-a998-c24804a8824b)

**Reply by feclaudia** [reply]
Maybe you can add this to your demo 2124? Thanks a lot!

**Comment by charlenni** [comment]
Is this more what you expected? Do you have example data that you could provide?

**Reply by feclaudia** [reply]
o.k. I have two examples: one polar and one Cartesian. The images show, what I want to reproduce. The projection has to be exact. The projection of the Cartesian data (900km x 900km) is decribed in this German document on pages 13-15: 
https://www.dwd.de/DE/leistungen/radolan/radolan_info/radolan_radvor_op_komposit_format_pdf.pdf?__blob=publicationFile&v=23

The polar data (360° x 128 km) have a point in lat/lon coordinates in the middle, which you can find in the file name.
I will work on this again in three weeks time. So you have time to look on my examples ;-) 

[examples.zip](https://github.com/Mapsui/Mapsui/files/12237225/examples.zip)

**Reply by charlenni** [reply]
Ok. Now I know what you try to do and what data you have.

One more question: Is this an open source or closed source project?

**Reply by charlenni** [reply]
Oh, some more things:
1. The raster of your TXT files is the same as described in the PDF document (X and Y direction, up and down)?
2. There is no table for connecting values with colors. Could you provide this too?

**Reply by feclaudia** [reply]
I gave you just the data matrix without the meta data, because parts are closed source. In the polar file each row is degree and each column distance from the middle. For the Cartesian I'm not sure about x+y...... it's not necessarily as described in the pdf. Maybe it is mirrored.
The polar data have a linear colour table (0-4-8-12.....) and the Cartesian have logarithmic colour table:
![scale](https://github.com/Mapsui/Mapsui/assets/102292078/fbe7a84f-16c5-47f4-93b0-850359f6a654)

**Reply by charlenni** [reply]
Any answer to open source or closed source?

**Comment by feclaudia** [comment]
I'm not sure, what you mean. The given examples are open source.

charlenni ***@***.***> schrieb am Mi., 2. Aug. 2023, 14:27:

**Reply by charlenni** [reply]
It makes much more fun to help someone , who is not making money with the help ;-)

**Comment by feclaudia** [comment]
Mmh o.k. now I understand..... If it works well, I will use it for work,
sorry! I'm not happy with my old routines for different reasons. So I
looked around what else I can use for mapping. If you stop helping now, I
can understand.... Thanks for your help so far!

charlenni ***@***.***> schrieb am Mi., 2. Aug. 2023, 16:07:

**Comment by charlenni** [comment]
Ok, I made a sample for your use case. I used an own style renderer to do this. First I create a SKPicture from the data and later show this picture on the screen (still as vector data). One problem was, that there are no square pixels. They are all polygons.

There are some problems:
1. Your color table for Cartesian colors doesn't belong to the data you provided
2. The picture isn't drawn on the correct place. It is a little bit to the west, but I didn't want to go too deep in the coordinate transformation.
3. There are still some performance gains ;-)

It now looks like
![grafik](https://github.com/Mapsui/Mapsui/assets/4381223/6f89f12b-8c9f-4a9f-945f-2f053ac19a41)
![grafik](https://github.com/Mapsui/Mapsui/assets/4381223/0a168d18-08d0-4540-93e2-e46605103ebb)
![grafik](https://github.com/Mapsui/Mapsui/assets/4381223/87ff31f1-4694-4074-9cd1-f77661f9d845)

I get on my system full map redraws under 200 ms and zoomed in redraws around 25 ms or faster.

You could find the sample [here](https://github.com/charlenni/RainRadar).

**Reply by feclaudia** [reply]
Hey! I'm back.... Thanks a lot for your example! I will test it soon and hope it is as fast as you wrote! That would be great :-)

**Reply by charlenni** [reply]
Hi, any news about this?

**Reply by feclaudia** [reply]
Hey! I didn't have time to test it for different reasons.... I hope, I have the time soon. Thanks for asking!

**Reply by feclaudia** [reply]
Finally I started today with your example and tried to understand, what you've done. Sorry for the big delay! 
So, if I get it right, you changed from using a bitmap to using SKPicture to have more possibilities with your own Style and StyleRenderer? I did't really get the thing with the 20 points ("// If we add more than 20 points, then there are rounding problems"). Does this concern, if you zoomed in or out of the picture?
I general I think, it it brilliant, that there is no need to draw the polygons over and over again during zooming. This is very time-saving! Good job!
If I want to load one image after another to create animations, I have to create always new SKPictures, right? 
Thanks a lot!

**Reply by charlenni** [reply]
The idea behind: Make an own StyleRenderer, so that you could divide rendering and drawing, what speeds up the drawing. You have static data, which could converted once to the Skia natives (in this case a SKPicture) and later be drawn by an own StyleRenderer. We have then a two-step process: first create the recorded drawing commands (slow because of the calculations) from the data and second draw this recorded drawing commands at the given position on the map (fast).

My first approach with only rect pixels wasn't possible, because the data isn't rectangular, but a circular cut-out. That was something you didn't mention before. 

The thing with the 20 points is, that I not draw circular cut-out for circular cut-out, but add a few of them together. That is faster. But I don't draw circular cut-outs, but trapezoids. It works for up to 20 trapezoids in a column, but then the difference to the original circular cut-outs gets too high. So add max. 20 data points with the same color in a column together and draw them at once.

If you want to have animations, then you are right: you have to create an SKPicture for each of the animation points and later decide, which to draw. You could create for each step an own feature with its own SKPicture and set this features to visible or anvisible over time. Or you create one feature with many SKPictures and decide while drawing, which SKPicture youn want to draw. Many possibilities.

**Reply by feclaudia** [reply]
o.k. and is there also a possibility to keep the path calculations in CreatePicture for the next image and just change the colours? Thanks!

**Reply by charlenni** [reply]
No, that isn't possible, because
1. Some parts are drawn together regarding the color, which could change
2. SKPicture records all the drawing including the used color, so it isn't possible to change this.

But the SKPictures are not very big in size. And the precalculation while loading them shouldn't take to much time. So create them once in the beginning and draw them later often.

**Reply by feclaudia** [reply]
Hey! I adapted the SKPicture method to my software. To get the right projection I saved my calculated polygon coordinates in data and stored the x/y min/max values for the rectangle in CreatePicture. So I can use the method as well for polar data. Moreover I separated the values from data, so that I can reuse the coordinates with new values without recalculating. Now I will test to handle many SKPictures.

**Reply by charlenni** [reply]
Thank you for keeping me informed.

One more thing? Which implementation of Mapsui do you use? WPF? Avalonia? Or something else?

**Reply by feclaudia** [reply]
I use WPF 4.0.

**Reply by charlenni** [reply]
I was afraid of that. I looked into this the last few days and found, that there is still no good GPU solution for WPF. So all you do on WPF is created by CPU without any GPU support. This is different on Android, iOS or WinForms :-(

**Reply by feclaudia** [reply]
o.k. I know nothing about CPU/GPU .... 
One other question: I tried to use my image export with Mapsui.Rendering.Skia.MapRenderer().RenderToBitmapStream, but it didn't export the new layer with the SKPicture. It does only export the other layers. All picture information are in Features. Do I have to change other attributes of the memorylayer? Thanks.

**Reply by feclaudia** [reply]
Hello! I still have this export problem. Any hint for me? Thanks

**Reply by charlenni** [reply]
Try to open a new issues, because I would say, this has nothing to do with original issue.
